### PR TITLE
Improve Data class

### PR DESCRIPTION
**Pull request by carlosggarcia**
Improved Data class:

- [x] Avoid redundancy (cleaner code)
- [x] Improve declaration of cls to compute: npz file, all by default and none by default.
- [x] Tests

### CODE REVIEW

**Comment by carlosggarcia** [comment]
@damonge  ready to review. I was thinking of changing some variable names (e.g `get_tracers_used` to `get_tracers`; `get_cl_trs_names` to `get_cl_tracers`, etc.) What do you think?

**Comment by carlosggarcia** [comment]
@damonge now it works. So feel free to review when you can. Also, check my last commit. I think it makes sense the change I've done to the test, but worth double checking.

File: xcell/cls/data.py [hunk 1]
```diff
@@ -40,9 +41,149 @@ def __init__(self, data_path='', data={}, override=False,
...
+
+    def _get_clcov_from_data_matrix(self, return_default=True):
+        conf = self._get_section('cov').get('cls_from_data', {})
+        # Backwards compatibility
+        if isinstance(conf, str):
+            combs = []
+            default = conf
+        elif isinstance(conf, list):
+            combs = conf
+            default = self.default_clcov_from_data
```
Comment (damonge) [inline xcell/cls/data.py:118]: this should be "all",  I think

File: xcell/cls/data.py [hunk 2]
```diff
@@ -40,9 +41,149 @@ def __init__(self, data_path='', data={}, override=False,
 
         os.makedirs(self.data['output'], exist_ok=True)
         self._check_yml_in_outdir(override, ignore_existing_yml)
-        self.tr_matrix = None
+        self.tracers = {'wsp': None, 'no_wsp': None}
```
Comment (damonge) [inline xcell/cls/data.py:44]: give this a different name

File: xcell/cls/data.py [hunk 3]
```diff
@@ -88,24 +229,25 @@ def read_data(self, data_path):
...
     def get_tracers_used(self, wsp=False):
-        tracers = []
-        for trk, trv in self.data['cls'].items():
-            tr1, tr2 = trk.split('-')
-            if trv['compute'] != 'None':
-                tracers.append(tr1)
-                tracers.append(tr2)
+        lab = ['no_wsp', 'wsp'][wsp]
+        if self.tracers[lab] is None:
+            tr_names = self._get_tracers_defined()
```
Comment (damonge) [inline xcell/cls/data.py:234]: Comment: first get all tracers in yaml file

File: xcell/cls/data.py [hunk 4]
```diff
@@ -88,24 +229,25 @@ def read_data(self, data_path):
...
+            tmat = self.get_tracer_matrix()
```
Comment (damonge) [inline xcell/cls/data.py:235]: comment: get matrix that tells you which cls you're computing

File: xcell/cls/data.py [hunk 5]
```diff
@@ -88,24 +229,25 @@ def read_data(self, data_path):
...
 
-        tracers_for_cl = []
-        for tr in self.data['tracers'].keys():
-            # Remove tracer number
-            tr_nn = self.get_tracer_bare_name(tr)
-            if tr_nn in tracers:
-                tracers_for_cl.append(tr)
+            # Saving only tr1 to keep order
```
Comment (damonge) [inline xcell/cls/data.py:237]: comment: `tracers` now will contain only the tracers for which we actually compute C_ells

File: xcell/cls/data.py [hunk 6]
```diff
@@ -88,24 +229,25 @@ def read_data(self, data_path):
...
+            tracers = []
+            for tr1 in tr_names:
+                for tr2 in tr_names:
+                    req = tmat[(tr1, tr2)]
+                    if req['compute'] and (tr1 not in tracers):
+                        tracers.append(tr1)
+                        break
+            if wsp:
+                tracers = self._filter_tracers_wsp(tracers)
```
Comment (damonge) [inline xcell/cls/data.py:246]: comment: we only keep those that have different masks

File: xcell/cls/data.py [hunk 7]
```diff
@@ -184,52 +275,34 @@ def get_tracers_bare_name_pair(self, tr1, tr2, connector='-'):
...
-                    pass
-                elif (clreq == 'auto') and (tr1 != tr2):
-                    continue
-                elif clreq == 'None':
-                    continue
-                cl_tracers.append((tr1, tr2))
-        return cl_tracers
-
     def get_cl_trs_names(self, wsp=False):
         lab = ['no_wsp', 'wsp'][wsp]
```
Comment (damonge) [inline xcell/cls/data.py:279]: add a short docstring just saying what this is meant to do. Doesn't have to be long.

File: xcell/cls/data.py [hunk 8]
```diff
@@ -184,52 +275,34 @@ def get_tracers_bare_name_pair(self, tr1, tr2, connector='-'):
...
+                        continue
+                    cl_tracers.append((tr1, tr2))
 
-        return cov_tracers
+            self.cl_tracers[lab] = cl_tracers
+
+        return self.cl_tracers[lab]
 
     def get_cov_trs_names(self, wsp=False):
         lab = ['no_wsp', 'wsp'][wsp]
```
Comment (damonge) [inline xcell/cls/data.py:297]: add docstring as for the previous method

**Review by damonge — CHANGES_REQUESTED** [review]
Haven't looked at the tests, but will do so next

File: xcell/cls/data.py (hunk 2 above)
Comment (carlosggarcia) [reply xcell/cls/data.py]: I was not very imaginative so I couldn't come up with anything better. I wrote a few comments though.

File: xcell/cls/data.py (hunk 3 above)
Comment (carlosggarcia) [reply xcell/cls/data.py:247]: Done

File: xcell/cls/data.py (hunk 6 above)
Comment (carlosggarcia) [reply xcell/cls/data.py:261]: Done

File: xcell/cls/data.py (hunk 7 above)
Comment (carlosggarcia) [reply xcell/cls/data.py]: Done

File: xcell/cls/data.py (hunk 8 above)
Comment (carlosggarcia) [reply xcell/cls/data.py]: Done

**Comment by carlosggarcia** [comment]
@damonge pinging you in case you didn't see the changes!

File: xcell/cls/data.py (hunk 4 above)
Comment (carlosggarcia) [reply xcell/cls/data.py:249]: This is done, I forgot to comment here

**Comment by carlosggarcia** [comment]
@damonge quick summary here too. I think it only remains that you have a look at how I addressed your comments and, if you're happy with them, accept and merge

File: xcell/cls/data.py (hunk 1 above)
Comment (damonge) [reply xcell/cls/data.py:130]: Not sure I understood, but sounds like you do

File: xcell/cls/data.py (hunk 5 above)
Comment (damonge) [reply xcell/cls/data.py:251]: I don't think you added this comment

**Review by damonge — COMMENTED** [review]
Just one thing

**Comment by carlosggarcia** [comment]
Not sure I understand. The comment is there, isn't it?

**Comment by carlosggarcia** [comment]
@damonge done.

**Review by damonge — APPROVED** [review]

**Pull request merged**